src/lib.rs

```diff
@@ -454,10 +454,31 @@ pub unsafe extern "C" fn bdd_pickcube(f: bdd_t) -> bdd_assignment_t {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn bdd_writebytes(f: bdd_t, path: *const std::ffi::c_char) -> () {
+pub unsafe extern "C" fn bdd_save(f: bdd_t, path: *const std::ffi::c_char) -> () {
     let f = unsafe { &**f._p };
     let f_bytes = f.to_bytes();
 
     let path_cstr = unsafe { std::ffi::CStr::from_ptr(path) };
     std::fs::write(Path::new(path_cstr.to_str().unwrap()), f_bytes).unwrap();
 }
+
+#[no_mangle]
+pub unsafe extern "C" fn bdd_load(manager: manager_t, path: *const std::ffi::c_char) -> bdd_t {
+    // Taken from Samuel Pastva and Thomas Henzinger's source code in 'util.rs' in their artifact:
+    // https://zenodo.org/records/7958052
+
+    let path_cstr = unsafe { std::ffi::CStr::from_ptr(path) };
+    let file = std::fs::File::open(path_cstr.to_str().unwrap()).unwrap();
+
+    // The size must be a multiple of 10 to circumvent this:
+    // https://github.com/sybila/biodivine-lib-bdd/issues/34
+    let mut file = std::io::BufReader::with_capacity(1024 * 10, file);
+    let mut f = Bdd::read_as_bytes(&mut file).unwrap();
+
+    // Ensure this `Bdd` is compatible with any other loaded using this very function (i.e. it
+    // tracks the same number of variables).
+    //assert!(f.num_vars() <= 65_535);
+    unsafe { f.set_num_vars(65_535); }
```

Comment:
To me, this looks somewhat hacky. Does your application rely on setting the domain this way or could we just remove it?

Reply:
This hack was part of Samuel Pastva's and Thomas Henzinger's paper. Though, they did set it to `10_000` only. But, to work with all inputs (also my hand-made ones), I had to increase it to the maximum value.

Reply:
If you remove this and run the Apply benchmark on the BDDs from Pastva and Henzinger's artifact, you end up in the following panic.
```
thread '<unnamed>' panicked at /home/soelvsten/.cargo/registry/src/index.crates.io-6f17d22bba15001f/biodivine-lib-bdd-0.5.19/src/_impl_bdd/_impl_boolean_ops.rs:247:9:
Var count mismatch: BDDs are not compatible. 258 != 102
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at library/core/src/panicking.rs:221:5:
panic in a function that cannot unwind
stack backtrace:
   0:           0x4986b5 - std::backtrace_rs::backtrace::libunwind::trace::h7549a4ccac42a6a7
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:           0x4986b5 - std::backtrace_rs::backtrace::trace_unsynchronized::h08410043b3f17d43
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:           0x4986b5 - std::sys::backtrace::_print_fmt::hd3d9cf94c019b9f5
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/sys/backtrace.rs:65:5
   3:           0x4986b5 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h37cba3c5f4c20c12
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/sys/backtrace.rs:40:26
   4:           0x4e598b - core::fmt::rt::Argument::fmt::h20a7bdc1ab81b185
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/fmt/rt.rs:173:76
   5:           0x4e598b - core::fmt::write::hc3ba8f02eab31f86
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/fmt/mod.rs:1182:21
   6:           0x48d55f - std::io::Write::write_fmt::h4486f8c78ef5f3b6
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/io/mod.rs:1827:15
   7:           0x49aea1 - std::sys::backtrace::BacktraceLock::print::h4bd29bd039770b0e
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/sys/backtrace.rs:43:9
   8:           0x49aea1 - std::panicking::default_hook::{{closure}}::h78630d6ff2d57eb0
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/panicking.rs:269:22
   9:           0x49ab7c - std::panicking::default_hook::h5ce7b558914edf7a
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/panicking.rs:296:9
  10:           0x49b5c1 - std::panicking::rust_panic_with_hook::h23abe07a80d694df
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/panicking.rs:800:13
  11:           0x49b3f3 - std::panicking::begin_panic_handler::{{closure}}::hd1358c96dd33f981
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/panicking.rs:667:13
  12:           0x498b79 - std::sys::backtrace::__rust_end_short_backtrace::h164f178ea4a25070
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/sys/backtrace.rs:168:18
  13:           0x49b0b4 - rust_begin_unwind
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/std/src/panicking.rs:665:5
  14:           0x40c5e5 - core::panicking::panic_nounwind_fmt::runtime::h5ef492a45ee08557
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/panicking.rs:112:18
  15:           0x40c5e5 - core::panicking::panic_nounwind_fmt::hda4beb2fdd57dbda
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/panicking.rs:122:5
  16:           0x40c672 - core::panicking::panic_nounwind::hda453532be4c4c75
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/panicking.rs:221:5
  17:           0x40c896 - core::panicking::panic_cannot_unwind::h291afb47df73dc3f
                               at /builddir/build/BUILD/rustc-1.81.0-src/library/core/src/panicking.rs:309:5
  18:           0x416ba3 - bdd_and
  19:           0x41002b - _ZZ9run_applyI18libbdd_bdd_adapterEiiPPcENKUlRS0_E_clES3_.isra.0
  20:           0x4121fe - _Z3runI18libbdd_bdd_adapterZ9run_applyIS0_EiiPPcEUlRS0_E_EiRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiRKT0_.constprop.0.isra.0
  21:           0x414322 - _Z9run_applyI18libbdd_bdd_adapterEiiPPc
  22:     0x7f24fb239088 - __libc_start_call_main
  23:     0x7f24fb23914b - __libc_start_main@@GLIBC_2.34
  24:           0x40ed25 - _start
  25:                0x0 - <unknown>
thread caused non-unwinding panic. aborting.
Aborted (core dumped)
```
It works fine for (a) the RelProd benchmark and (b) my hand-made examples for the Apply benchmark because (a) they were created by the same program (and hence dealt with the same number of variables) and (b) were hand-made to be using the maximum number of variables.

Reply:
Ok, fair enough. I’m wondering whether it would be preferable to pass the number of variables as a parameter, but we can merge it as is for now.